File: nightgrass/src/main/java/com/goodfeel/nightgrass/serviceImpl/ProductService.java
package com.goodfeel.nightgrass.serviceImpl;

import com.goodfeel.nightgrass.data.Product;
import com.goodfeel.nightgrass.repo.ProductRepository;
import com.goodfeel.nightgrass.service.IProductService;
import com.goodfeel.nightgrass.dto.ProductDto;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
public class ProductService implements IProductService {

    private final ProductRepository productRepository;

    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    @Override
    public Flux<ProductDto> getAllProducts() {

        Flux<Product> products = productRepository.findAll();
        return products.map(product ->
                new ProductDto(
                        product.getId(),
                        product.getName(), product.getDescription(), product.getImageUrl(), product.getPrice())
        );
    }

    @Override
    public Mono<ProductDto> getProductById(Long id) {
        return productRepository.findById(id).map(product -> {
            return new ProductDto(product.getId(), product.getName(),
                    product.getDescription(), product.getImageUrl(), product.getPrice());
        });
    }

    @Override
    public Mono<Product> saveProduct(Product product) {
        return productRepository.save(product);
    }

    @Override
    public Mono<Void> deleteProduct(Long id) {
        return productRepository.deleteById(id);
    }

}

File: nightgrass/src/main/kotlin/com/goodfeel/nightgrass/serviceImpl/ProductService.kt
package com.goodfeel.nightgrass.serviceImpl

import com.goodfeel.nightgrass.data.Product
import com.goodfeel.nightgrass.dto.ProductDto
import com.goodfeel.nightgrass.exception.ForeignKeyConstraintViolationException
import com.goodfeel.nightgrass.repo.ProductRepository
import com.goodfeel.nightgrass.service.IProductService
import com.goodfeel.nightgrass.service.ProcessedProductService
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class ProductService(
    private val productRepository: ProductRepository,
    private val processedProductService: ProcessedProductService
) : IProductService {

    override fun allProducts(): Flux<ProductDto> {
        return processedProductService.findAllAndProcessProducts()
            .onErrorResume { ex ->
                Flux.error(RuntimeException("Failed to fetch all products: ${ex.message}", ex))
            }
    }

    override fun getProductById(id: Long): Mono<ProductDto> {
        return processedProductService.findAndProcessProductByProductId(id)
            .onErrorResume { ex ->
                Mono.error(RuntimeException("Error fetching product with ID $id: ${ex.message}", ex))
            }
            .switchIfEmpty(Mono.error(NoSuchElementException("Product with ID $id not found")))

    }

    override fun saveProduct(product: Product): Mono<Product> {
        return productRepository.save(product)
            .onErrorResume { ex ->
                Mono.error(RuntimeException("Failed to save product: ${ex.message}", ex))
            }
    }

    override fun deleteProduct(id: Long): Mono<Void> {
        return productRepository.deleteById(id)
            .onErrorMap { ex ->
                if (ex.message?.contains("foreign key") == true) {
                    ForeignKeyConstraintViolationException(ex.message!!)
                } else {
                    RuntimeException("Failed to delete product with ID $id: ${ex.message}", ex)
                }

            }
    }

    override fun getTop3BigHits(): Flux<ProductDto> {
        return processedProductService.findTop3BigHits()
    }

    override fun getTop8PopularOrNewProducts(): Flux<ProductDto> =
        processedProductService.findTop8PopularOrNewProducts()

}
